test/clickhousex/http_client_test.exs

```diff
@@ -0,0 +1,138 @@
+      :ok = Bypass.down(bypass)
+      assert {:ok, conn} = Client.disconnect(conn)
+      assert conn.state == :closed
+    end
+
+    test "when connection is open", %{conn: conn} do
+      assert {:ok, conn} = Client.disconnect(conn)
+      assert conn.state == :closed
+    end
+  end
+
+  describe "request/7" do
+    setup :connected
+
+    # curl 'localhost:8123/?database=no_db&query=SELECT+1' -d " FORMAT JSONCompact" -v
+    test "when server responds with error", %{conn: conn, bypass: bypass} do
+      Bypass.expect_once(bypass, "POST", "/", fn conn ->
+        assert conn.query_string == "database=no_db&query=SELECT+1"
+        {:ok, " FORMAT JSONCompact", conn} = Plug.Conn.read_body(conn)
+
+        Plug.Conn.send_resp(
+          conn,
+          404,
+          "Code: 81. DB::Exception: Database no_db doesn't exist. (UNKNOWN_DATABASE) (version 22.7.1.2163 (official build))"
+        )
+      end)
+
+      query = Query.new("SELECT 1")
+      request = DBConnection.Query.encode(query, _params = [], _opts = [])
+
+      assert {:error, _conn, error} =
+               Client.request(
+                 conn,
+                 query,
+                 request,
+                 Clickhousex.timeout(),
+                 _username = nil,
+                 _password = nil,
+                 _database = "no_db"
+               )
+
+      assert error ==
+               "Code: 81. DB::Exception: Database no_db doesn't exist. (UNKNOWN_DATABASE) (version 22.7.1.2163 (official build))"
+    end
+
+    # curl 'localhost:8123/?database=default&query=SELECT+1' -d " FORMAT JSONCompact" -v
+    test "when server responds with 200 to `SELECT 1`", %{conn: conn, bypass: bypass} do
+      Bypass.expect_once(bypass, "POST", "/", fn conn ->
+        assert conn.query_string == "database=default&query=SELECT+1"
+        {:ok, " FORMAT JSONCompact", conn} = Plug.Conn.read_body(conn)
+
+        resp = """
+        {
+          "meta":
+          [
+            {
+              "name": "1",
+              "type": "UInt8"
+            }
+          ],
+
+          "data":
+          [
+            [1]
+          ],
+
+          "rows": 1,
+
+          "statistics":
+          {
+            "elapsed": 0.000349,
+            "rows_read": 1,
+            "bytes_read": 1
+          }
+        }
+        """
+
+        conn
+        |> Plug.Conn.put_resp_content_type("application/json")
+        |> Plug.Conn.send_resp(200, resp)
+      end)
+
+      query = Query.new("SELECT 1")
+      request = DBConnection.Query.encode(query, _params = [], _opts = [])
+
+      assert {:ok, %Mint.HTTP1{state: :open}, result} =
+               Client.request(
+                 conn,
+                 query,
+                 request,
+                 Clickhousex.timeout(),
+                 _username = nil,
+                 _password = nil,
+                 _database = "default"
+               )
+
+      assert result == {:selected, ["1"], [[1]]}
+    end
```

Comment:
@ukutaht These are more like drafts right now, just wanted to get feedback before I continue adding more tests.

Reply:
I plan to cover every major branch in the http_client.ex logic. And then maybe refactor to make the logic under test more apparent. Like:

```
test "when server responds with 200 to `SELECT 1`", %{conn: conn, bypass: bypass} do
  bottled_response(bypass, :no_db)
  assert {:error, _conn, error} = request(conn, "SELECT 1")
  assert error == "Code: 81. DB::Exception: Database no_db doesn't exist. (UNKNOWN_DATABASE) ( ..."
end
```

Reply:
Plan looks good to me!